components/director/internal/systemfetcher/service.go

```diff
@@ -0,0 +1,195 @@
+package systemfetcher
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/kyma-incubator/compass/components/director/internal/domain/tenant"
+	"github.com/kyma-incubator/compass/components/director/internal/model"
+	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+	"github.com/kyma-incubator/compass/components/director/pkg/persistence"
+	"github.com/pkg/errors"
+)
+
+//go:generate mockery --name=TenantService --output=automock --outpkg=automock --case=underscore
+type TenantService interface {
+	List(ctx context.Context) ([]*model.BusinessTenantMapping, error)
+	GetInternalTenant(ctx context.Context, externalTenant string) (string, error)
+}
+
+//go:generate mockery --name=SystemsService --output=automock --outpkg=automock --case=underscore
+type SystemsService interface {
+	CreateManyIfNotExistsWithEventualTemplate(ctx context.Context, applicationInputs []model.ApplicationRegisterInput, systemToTemplateMapping []string) error
+}
+
+//go:generate mockery --name=ApplicationTemplateService --output=automock --outpkg=automock --case=underscore
+type ApplicationTemplateService interface {
+	GetByName(ctx context.Context, name string) (*model.ApplicationTemplate, error)
+}
+
+//go:generate mockery --name=SystemsAPIClient --output=automock --outpkg=automock --case=underscore
+type SystemsAPIClient interface {
+	FetchSystemsForTenant(ctx context.Context, tenant string) ([]System, error)
+}
+
+type SystemFetcher struct {
+	transaction        persistence.Transactioner
+	tenantService      TenantService
+	systemsService     SystemsService
+	systemsAPIClient   SystemsAPIClient
+	appTemplateService ApplicationTemplateService
+
+	workers chan struct{}
+}
+
+func NewSystemFetcher(tx persistence.Transactioner, ts TenantService, ss SystemsService, sac SystemsAPIClient, appTemplateService ApplicationTemplateService, fetcherParallellism int) *SystemFetcher {
+	return &SystemFetcher{
+		transaction:        tx,
+		tenantService:      ts,
+		systemsService:     ss,
+		systemsAPIClient:   sac,
+		appTemplateService: appTemplateService,
+		workers:            make(chan struct{}, fetcherParallellism),
+	}
+}
+
+type TenantSystems struct {
+	tenant  *model.BusinessTenantMapping
+	systems []System
+}
+
+func (s *SystemFetcher) SyncSystems(ctx context.Context) error {
+	tenants, err := s.listTenants(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to list tenants")
+	}
+
+	systemsQueue := make(chan TenantSystems, 100)
+	wgDB := sync.WaitGroup{}
+	wgDB.Add(1)
+	go func() {
+		defer func() {
+			wgDB.Done()
+		}()
+		for tenantSystems := range systemsQueue {
+			err = s.saveSystemsForTenant(ctx, tenantSystems.tenant, tenantSystems.systems)
```

Comment:
The `saveSystemsForTenant` opens a transaction inside. Is it thread-safe to use `s.transaction.Begin()` multiple times by different goroutines? If yes, then does every one of these transactions open up a new db connection, because if it is the case, this could exhaust our connections

Reply:
You are correct to some extent. It does start a new db connection and can possibly exhaust them, but it actually does not run multiple times in parallel. So to say it should be safe to use as long as you don't run it in parallel.

The new goroutine is only running one `saveSystemsForTenant` at a time, but it does start a new connection each time after it has finished the previous processing loop and closed(commited/rolled back) the previous connection.